sw-emulator/lib/cpu/src/cpu.rs

```diff
@@ -262,7 +339,27 @@ impl<TBus: Bus> Cpu<TBus> {
     ///
     /// * `RvException` - Exception with cause `RvExceptionCause::IllegalRegister`
     pub fn write_xreg(&mut self, reg: XReg, val: RvData) -> Result<(), RvException> {
-        self.xregs.write(reg, val)
+        // XReg::X2 is the sp register.
+        if reg == XReg::X2 {
+            self.check_stack(val);
+        }
+        self.xregs.write(reg, val)?;
+        Ok(())
+    }
+
+    // Check if the stack overflows at the requested address.
+    //
+    // Panics if the stack will overflow.
+    fn check_stack(&mut self, val: RvData) {
+        if let Some(stack_info) = &mut self.stack_info {
+            if stack_info.check_overflow(val) {
+                panic!(
```

Comment:
Alternatively, this could be a print statement and the panic ocurring in `Cpu::drop`.

This would allow the max stack overflow to be observed.

Reply:
I'm making this change. It makes it a lot easier to root cause stack regressions.